### PR TITLE
Add FullRecorder for logging an entire optimization

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
I'm writing some personal code now to verify.

I'm not sure "Read" is the function name. Maybe Read/RealAll as a parallel to csv?

**Comment by btracey** [comment]
If we do change to Read/ReadAll, do we change the struct to also have an io.Reader? Do we make a different type entirely?

**Comment by btracey** [comment]
This doesn't record the initial optimization step.

File: recorder.go [hunk 1]
```diff
@@ -104,3 +105,72 @@ func (p *Printer) Record(loc *Location, op Operation, stats *Stats) error {
 	p.lastValue = time.Now()
 	return nil
 }
+
+// Full Recorder records all of the evaluations that occur during an optimization
```
Comment (vladimir-ch) [inline recorder.go]: Do you have a justification why this should be in optimize? The functionality seems quite specific to fit general needs.

Comment (btracey) [inline recorder.go]: It seems specific? In what way? It seems like a natural counterpart to Printer. Printer prints to a writer to track optimization, this saves the optimization run to `json` for retrospection

Comment (vladimir-ch) [inline recorder.go]: I was thinking about what is being recorded and into what format. I can imagine that either will easily not fit the needs and the user will write it's own recorder.

File: recorder.go [hunk 2]
```diff
@@ -104,3 +105,72 @@ func (p *Printer) Record(loc *Location, op Operation, stats *Stats) error {
...
+
+func (f FullRecorder) Record(loc *Location, op Operation, stats *Stats) error {
+	r := RecordLocation{Op: op.String()}
+	switch {
+	case op.isEvaluation():
+		r.Loc.X = loc.X
+		switch op {
+		default:
+			panic("optimize: unknown evaluation operation")
+		case FuncEvaluation:
```
Comment (vladimir-ch) [inline recorder.go]: Evaluation operations can be ORed together which this switch does not take into account.

File: recorder.go [hunk 3]
```diff
@@ -104,3 +105,72 @@ func (p *Printer) Record(loc *Location, op Operation, stats *Stats) error {
...
+			r.Loc.F = loc.F
+		case GradEvaluation:
+			r.Loc.Gradient = loc.Gradient
+		case HessEvaluation:
+			r.Loc.Hessian = loc.Hessian
+		}
+	case op == MajorIteration || op == PostIteration:
+		r.Loc = *loc
+	}
+	if op.isEvaluation() {
```
Comment (vladimir-ch) [inline recorder.go]: A left-over before the above switch was introduced?

File: recorder.go [hunk 4]
```diff
@@ -104,3 +105,72 @@ func (p *Printer) Record(loc *Location, op Operation, stats *Stats) error {
...
+	}
+	b, err := json.MarshalIndent(r, "", "\t")
+	if err != nil {
+		return nil
+	}
+	_, err = f.Writer.Write(b)
+	return err
+}
+
+// Read reads the stream written to by FullRecorder, and returns the history of
```
Comment (vladimir-ch) [inline recorder.go]: `written to r` ?

File: recorder.go (hunk 1 above)
Comment (vladimir-ch) [inline recorder.go]: `FullRecorder`

File: recorder.go [hunk 5]
```diff
@@ -104,3 +105,72 @@ func (p *Printer) Record(loc *Location, op Operation, stats *Stats) error {
 	p.lastValue = time.Now()
 	return nil
 }
+
+// Full Recorder records all of the evaluations that occur during an optimization
+// run. If Operation is an Evaulation, FullRecorder records the x location and
+// the corresponding field of Location. If Operation is a MajorIteration, or
```
Comment (vladimir-ch) [inline recorder.go]: field or fields

File: recorder.go (hunk 1 above)
Comment (btracey) [inline recorder.go]: I understand your point about things being Ored and this not handling that properly. Do you have other suggestions for making it more general? It seems reasonable to me to provide some built-in way to store the history of the optimization for retrospection (in addition to the human-friendly printing we have now).

**Comment by btracey** [comment]
Okay, I think I see your point about Json being speciifc. Maybe the full recorder can just keep a list of the steps? That was the user can do something with them in memory, or save them using Json after the optimization has concluded (or whatever else).

**Comment by vladimir-ch** [comment]
Closing because this repository is no longer maintained. Development has moved to https://github.com/gonum/gonum.

**Pull request closed without merging**